### ISSUE TITLE
Issues & Suggestions

### ISSUE THREAD

**Issue by halimpuckjava** [edvin/tornadofx#1034]
hi

1. i try to get scene from view in init() bloc but always it return null, how to fix this
2. interpolation for transition
3. lifecyle for view & fragment & application

**Comment by edvin** [comment]
1. Right after init(), the framework has not had a chance to connect your UIComponent to a Scene yet, so you should do whatever you need to do in `onDock` instead.

2. What is it that's hard to do without it today? JavaFX already have start/stop events for Application

3. 
a) Please provide a usage example for what you're trying to achieve
b) That's easily done by using `find` first, and I feel the call interface to `replaceWith` would be a bit muddy if we add it there as well.

4. Could `FadeThrough` be what you need perhaps?

**Comment by halimpuckjava** [comment]
thanks

for point 3.a i have simple case: 

i have 2 views listUser & editUser (contains a FAB edit button), so when switch from listUser to editUser i want to animate (show with scale transtion) FAB button when this switch transition complete, to do this in javafx classic way i just handle this with setOnFinish(action: () -> Unit)

**Comment by edvin** [comment]
Aha! Could you try to animate that button in onDock? Not sure if it will work, but worth a shot. @ruckustboom - what do you think about adding an onFinish callback to the transitions?

**Comment by halimpuckjava** [comment]
yes its work with onDock() callback , just i want to add onFinish callback when transition complete

i have a other question:
how to implement PreloadApplication with torrnedofx ??

**Comment by edvin** [comment]
There is nothing special about TornadoFX with regards to the `Preloader` functionality in JavaFX, it would work like any vanilla JavaFX app.

`onCreate` is a callback for the `Workbench` actions, and has nothing to do with the life cycle of the component itself. This is described in the guide I believe :)

**Comment by halimpuckjava** [comment]
maybe i dont see, i suggest to add more docs about view (fragement) lifecycle to make this callback more clear 

thanks

**Comment by ruckustboom** [comment]
@edvin That shouldn't be too hard. We can just add an optional parameter to `replaceWith`. Would it make more sense to have it be `() -> Unit` or `(oldView, newView) -> Unit`?

I'm not sure I can really test this out, as I can no longer run JDK 8, but it should be pretty straight forward to implement.

**Comment by ruckustboom** [comment]
The choice between `() -> Unit` and `(oldView, newView) -> Unit` has a notable wrinkle. If we choose the latter, we then have to discuss how to handle both functions, as we currently have `replaceWith` for `Node`s as well as `UIComponent`s (the latter simple forwards to the former using the `root`s). We'd obviously have to do `(oldView: Node, newView: Node) -> Unit` for the `Node` version, but would we use the same for the `UIComponent` version, or use `(oldView: UIComponent, newView: UIComponent) -> Unit`?

Personally, I think we should just use `() -> Unit` and avoid the issue altogether.

**Comment by edvin** [comment]
@halimpuckjava about docs - the `onCreate` function is quite clearly documented in the javadoc, so I think that's good enough covered. If you want to contribute something to the guide you're most welcome.

@ruckustboom I'm torn, but I believe `() -> Unit` would achieve what he wants to do, and not muddy the waters too much otherwise :)

**Comment by ruckustboom** [comment]
[Here's a patch that should do the job](https://github.com/edvin/tornadofx/files/3428078/add_after_callback_to_view_transition.patch.txt), but I can't test it.

I'm not sure how I feel about the name, but `onFinished` and `onComplete` would both create conflicts, so I just grabbed something. You may want to rename it.
